### PR TITLE
:bug: :mag: When I search for an item I should only see relevant results fix #466

### LINKED ISSUE #466: 🐛 🔍 When I search for an item I should only see relevant results
**Technology and Knowledge Scope 🧠**

- [x] HTML, SCSS
- [x] Typescript

**Describe the bug**
When I search for a specific item, the unrelated blocks are listed as well not just what I'm looking for. 

**To Reproduce**
1. Go to a story overview page.
2. Click on the search bar and type a search term e.g image
4. See error.

**Expected behavior**
When I search for an element, I should only see results related to that specific search and not the unrelated blocks.

**Screenshots**
search bug

### PR DESCRIPTION
**Pull request by Beulah-Matt**
# Description
Before, when a user searched for a block, the block was filtered but the headings of other blocks were also showing below the relevant result. 
For this fix, I am suggesting that for a filter to fully match, both the heading and block texts have to match, thus returning only the relevant results for both the category and blocks under it.

Fixes #466 

## Type of change
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

# Screenshot (optional
[Screencast from 28-06-23 16:18:58.webm](https://github.com/italanta/elewa/assets/109944021/01ea5ce1-5bf3-4c62-bb4a-bff27070aefb)
)

# How Has This Been Tested?
Only relevant results show on search.

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: libs/features/convs-mgr/stories/editor/src/lib/components/blocks-library/blocks-library.component.ts [hunk 1]
```diff
@@ -185,13 +186,31 @@ export class BlocksLibraryComponent implements OnInit, OnDestroy {
...
+    .pipe(map(([filter, blocksArray]) => {
+      const filteredBlocks = blocksArray.filter((block: StoryBlock) => {
+        return block.message?.toString().toLowerCase().includes(filter);
+      });
+
+      // Check if any of the filtered blocks have a message that matches the filter
+      const hasMatchingMessage = filteredBlocks.some((block: StoryBlock) => {
+        return block.message?.toString().toLowerCase().includes(filter);
+      });
+
```
Comment (LemmyMwaura) [inline libs/features/convs-mgr/stories/editor/src/lib/components/blocks-library/blocks-library.component.ts:198]: No need for this as you can just check whether the filtered Block Property above is of length > 0

File: libs/features/convs-mgr/stories/editor/src/lib/components/blocks-library/blocks-library.component.ts [hunk 2]
```diff
@@ -185,13 +186,31 @@ export class BlocksLibraryComponent implements OnInit, OnDestroy {
...
+      // Set inputText to true if there is at least one matching message
+      this.inputText = hasMatchingMessage;
+
```
Comment (LemmyMwaura) [inline libs/features/convs-mgr/stories/editor/src/lib/components/blocks-library/blocks-library.component.ts:201]: I don't think this variable is necessary

**Review by LemmyMwaura — CHANGES_REQUESTED** [review]
Good Attempt so far but this solution will lead to performance issues as Ngif will constantly keep calling the hasMatchingBlocks method every second as it keeps track of the value before deciding whether to show an item. 

To achieve our desired output we'll need to do all the filtering in the ts file and store the value/values in a variable or list that will allow us to decide whether to render that item.

File: libs/features/convs-mgr/stories/editor/src/lib/components/blocks-library/blocks-library.component.ts (hunk 2 above)
Comment (Beulah-Matt) [reply libs/features/convs-mgr/stories/editor/src/lib/components/blocks-library/blocks-library.component.ts:201]: @LemmyMwaura I am working on this. I'll let you know once I am done.

**Review by LemmyMwaura — CHANGES_REQUESTED** [review]
This is a much better approach but we're still making a function call in out template to evaluate an ngIf condition. Challenge: Use a pipe to determine whether we should render the title, The pipe should only listen to changes when the user types on the search field (if you need any help implementing this reach out

**Comment by Beulah-Matt** [comment]
@LemmyMwaura , I have made some changes, let me know if they are what you had in mind.

**Review by LemmyMwaura — APPROVED** [review]
Yeah, this is perfect Well done 👏 
Made a few changes just to optimise it and improve the structure.

**Pull request merged**